File: Day27/PathSum.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def hasPathSum(self, root, targetSum):
        if not root:
            return False
        if not root.left and not root.right:
            return root.val == targetSum
        left = self.hasPathSum(root.left, targetSum - root.val)
        right = self.hasPathSum(root.right, targetSum - root.val)
        return left or right

def build_tree(nodes):
    if not nodes or nodes[0] == 'null':
        return None
    root = TreeNode(int(nodes[0]))
    queue = [root]
    i = 1
    while queue and i < len(nodes):
        current = queue.pop(0)
        if nodes[i] != 'null':
            current.left = TreeNode(int(nodes[i]))
            queue.append(current.left)
        i += 1
        if i < len(nodes) and nodes[i] != 'null':
            current.right = TreeNode(int(nodes[i]))
            queue.append(current.right)
        i += 1
    return root

if __name__ == "__main__":
    node_str = input("Enter tree nodes in level order (use 'null' for missing):\n")
    nodes = node_str.strip().split()
    root = build_tree(nodes)
    targetSum = int(input("Enter target sum:\n"))
    sol = Solution()
    print("Path Sum Exists?" , sol.hasPathSum(root, targetSum))

File: Day27/PathSum.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) { this.val = val; }
}

class PathSum {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        if (root.left == null && root.right == null) return root.val == targetSum;
        return hasPathSum(root.left, targetSum - root.val) || hasPathSum(root.right, targetSum - root.val);
    }

    public static TreeNode buildTree(String[] nodes) {
        if (nodes.length == 0 || nodes[0].equals("null")) return null;
        TreeNode root = new TreeNode(Integer.parseInt(nodes[0]));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;
        while (queue.size() > 0 && i < nodes.length) {
            TreeNode current = queue.poll();
            if (i < nodes.length && !nodes[i].equals("null")) {
                current.left = new TreeNode(Integer.parseInt(nodes[i]));
                queue.add(current.left);
            }
            i++;
            if (i < nodes.length && !nodes[i].equals("null")) {
                current.right = new TreeNode(Integer.parseInt(nodes[i]));
                queue.add(current.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter tree nodes in level order (use 'null' for missing):");
        String[] nodes = sc.nextLine().split(" ");
        TreeNode root = buildTree(nodes);
        System.out.println("Enter target sum:");
        int targetSum = sc.nextInt();
        PathSum sol = new PathSum();
        System.out.println("Path Sum Exists? " + sol.hasPathSum(root, targetSum));
        sc.close();
    }
}
